rllib/utils/replay_buffers/episode_replay_buffer.py

```diff
@@ -0,0 +1,508 @@
+    trajectory of length T. Each trajectory contains consecutive timesteps from an
+    episode, but might not start at the beginning of that episode. Should an episode end
+    within such a trajectory, a random next episode (starting from its t0) will be
+    concatenated to that "row". Example: `sample(B=2, T=4)` ->
+
+       0 .. 1 .. 2 .. 3  <- T-axis
+    0 e5   e6   e7   e8
+    1 f2   f3   h0   h2
+    ^ B-axis
+
+    .. where e, f, and h are different (randomly picked) episodes, the 0-index (e.g. h0)
+    indicates the start of an episode, and `f3` is an episode end (gym environment
+    returned terminated=True or truncated=True).
+
+    0-indexed returned timesteps contain the reset observation, a dummy 0.0 reward, as
+    well as the first action taken in the episode (action picked after observing
+    obs(0)).
+    The last index in an episode (e.g. f3 in the example above) contains the final
+    observation of the episode, the final reward received, a dummy 0-action, as well
+    as either terminated=True or truncated=True.
+    """
+    def __init__(
+        self,
+        capacity: int = 10000,
+        *,
+        batch_size_B: int = 16,
+        batch_length_T: int = 64,
+    ):
+        """Initializes an EpisodeReplayBuffer instance.
+
+        Args:
+            capacity: The total number of timesteps to be storable in this buffer.
+                Will start ejecting old episodes once this limit is reached.
+            batch_size_B: The number of rows in a SampleBatch returned from `sample()`.
+            batch_length_T: The length of each row in a SampleBatch returned from
+                `sample()`.
+        """
+        self.capacity = capacity
+        self.batch_size_B = batch_size_B
+        self.batch_length_T = batch_length_T
+
+        # The actual episode buffer.
+        self.episodes = deque()
+        # Maps (unique) episode IDs to the index under which to find this episode
+        # within our `self.episodes` deque. Note that after ejection started, the
+        # indices will NOT be changed. We will therefore need to offset these indices
+        # by the number of episodes that have already been ejected.
+        self.episode_id_to_index = {}
+        # The number of episodes that have already been ejected from the buffer
+        # due to reaching capacity. This is the offset, which we have to subtract
+        # from the episode index to get the actual location within `self.episodes`.
+        self._num_episodes_ejected = 0
+
+        # List storing all index tuples: [eps_idx, ts_in_eps_idx], where ...
+        # `eps_idx - self._num_episodes_ejected' is the index into self.episodes.
+        # `ts_in_eps_idx` is the timestep index within that episode
+        #  (0 = 1st timestep, etc..).
+        # We sample uniformly from the set of these indices in a `sample()`
+        # call.
+        self._indices = []
+
+        # The size of the buffer in timesteps.
+        self.size = 0
```

Comment:
Can we consolidate this with
```
self._num_timesteps_added
self._num_timesteps_added_wrap
```
which are used in other Replay Buffers?
The exact name is not important to me but I think it should start with an underscore and include the phrase "ts" or "timesteps".
All buffers should carry this property since we have to limit size somehow in any case.

Reply:
Hmm, these are not really the same though.

`self.size` is the current size of the buffer, which is different from the number of ts added thus far (at last past the eviction point).

But I'm happy to rename for clarification purposes as well as add these two stats to the properties.

Reply:
fixed